core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -368,35 +407,44 @@ public void onFailure(Throwable e) {
-            if (primary.active() == false) {
-                logger.trace("primary shard [{}] is not yet active, scheduling a retry.", primary.shardId());
-                retryBecauseUnavailable(shardIt.shardId(), "Primary shard is not active or isn't assigned to a known node.");
+            if (state.nodes().nodeExists(primary.currentNodeId()) == false) {
+                logger.trace("primary shard [{}] is assigned to an unknown node [{}], scheduling a retry: action [{}], request [{}], cluster state version [{}]", request.shardId(), primary.currentNodeId(), actionName, request, state.version());
+                retryBecauseUnavailable(request.shardId(), "primary shard isn't assigned to a known node.");
                 return;
             }
-            if (observer.observedState().nodes().nodeExists(primary.currentNodeId()) == false) {
-                logger.trace("primary shard [{}] is assigned to anode we do not know the node, scheduling a retry.", primary.shardId(), primary.currentNodeId());
-                retryBecauseUnavailable(shardIt.shardId(), "Primary shard is not active or isn't assigned to a known node.");
-                return;
+            if (primary.currentNodeId().equals(state.nodes().localNodeId())) {
+                if (logger.isTraceEnabled()) {
+                    logger.trace("send action [{}] on primary [{}] for request [{}] with cluster state version [{}] to [{}] ", transportPrimaryAction, request.shardId(), request, state.version(), primary.currentNodeId());
+                }
+                performAction(primary.currentNodeId(), transportPrimaryAction, true);
+            } else {
+                if (logger.isTraceEnabled()) {
+                    logger.trace("send action [{}] on primary [{}] for request [{}] with cluster state version [{} to [{}]]", actionName, request.shardId(), request, state.version(), primary.currentNodeId());
+                }
+                performAction(primary.currentNodeId(), actionName, false);
             }
-            routeRequestOrPerformLocally(primary, shardIt);
         }
 
         /**
-         * checks for any cluster state blocks. Returns true if operation is OK to proceeded.
+         * checks for any cluster state blocks. Returns true if operation is OK to proceed and request is resolved
          * if false is return, no further action is needed. The method takes care of any continuation, by either
-         * responding to the listener or scheduling a retry
+         * responding to the listener or scheduling a retry.
          */
-        protected boolean checkBlocks() {
-            ClusterBlockException blockException = checkGlobalBlock(observer.observedState());
+        protected boolean checkBlocksAndResolveRequest() {
+            final ClusterState clusterState = observer.observedState();
+            ClusterBlockException blockException = clusterState.blocks().globalBlockedException(globalBlockLevel());
             if (blockException != null) {
```

Comment:
The handling here is duplicated in the handling after checking the index blocks. Can we collapse this logic?